src/_P090_MitsubishiHeatPump .ino

```diff
@@ -0,0 +1,698 @@
+              break;
+            }
+          }
+        }
+
+        if(header[1] == 0x61) { //Last update was successful
+          return RCVD_PKT_UPDATE_SUCCESS;
+        } else if(header[1] == 0x7a) { //Last update was successful
+          _isConnected = true;
+          return RCVD_PKT_CONNECT_SUCCESS;
+        }
+      }
+    }
+  }
+
+  return RCVD_PKT_FAIL;
+}
+
+void P090_data_struct::createInfoPacket(byte *packet, byte packetType) {
+  // add the header to the packet
+  for (int i = 0; i < INFOHEADER_LEN; i++) {
+    packet[i] = INFOHEADER[i];
+  }
+
+  // set the mode - settings or room temperature
+  if(packetType != PACKET_TYPE_DEFAULT) {
+    packet[5] = INFOMODE[packetType];
+  } else {
+    // request current infoMode, and increment for the next request
+    packet[5] = INFOMODE[_infoMode];
+    if(_infoMode == (INFOMODE_LEN - 1)) {
+      _infoMode = 0;
+    } else {
+      _infoMode++;
+    }
+  }
+
+  // pad the packet out
+  for (int i = 0; i < 15; i++) {
+    packet[i + 6] = 0x00;
+  }
+
+  // add the checksum
+  byte chkSum = checkSum(packet, 21);
+  packet[21] = chkSum;
+}
+
+void P090_data_struct::createPacket(byte *packet, const heatpumpSettings& settings) {
+  //preset all bytes to 0x00
+  for (int i = 0; i < 21; i++) {
+    packet[i] = 0x00;
+  }
+  for (int i = 0; i < HEADER_LEN; i++) {
+    packet[i] = HEADER[i];
+  }
+  if(settings.power != _currentSettings.power) {
+    packet[8]  = POWER[lookupByteMapIndex(POWER_MAP, 2, settings.power)];
+    packet[6] += CONTROL_PACKET_1[0];
+  }
+  if(settings.mode!= _currentSettings.mode) {
+    packet[9]  = MODE[lookupByteMapIndex(MODE_MAP, 5, settings.mode)];
+    packet[6] += CONTROL_PACKET_1[1];
+  }
+  if(!_tempMode && settings.temperature!= _currentSettings.temperature) {
+    packet[10] = TEMP[lookupByteMapIndex(TEMP_MAP, 16, settings.temperature)];
+    packet[6] += CONTROL_PACKET_1[2];
+  }
+  else if(_tempMode && settings.temperature!= _currentSettings.temperature) {
+    float temp = (settings.temperature * 2) + 128;
+    packet[19] = (int)temp;
+    packet[6] += CONTROL_PACKET_1[2];
+  }
+  if(settings.fan!= _currentSettings.fan) {
+    packet[11] = FAN[lookupByteMapIndex(FAN_MAP, 6, settings.fan)];
+    packet[6] += CONTROL_PACKET_1[3];
+  }
+  if(settings.vane!= _currentSettings.vane) {
+    packet[12] = VANE[lookupByteMapIndex(VANE_MAP, 7, settings.vane)] | (_wideVaneAdj ? 0x80 : 0x00);
+    packet[6] += CONTROL_PACKET_1[4];
+  }
+  if(settings.wideVane!= _currentSettings.wideVane) {
+    packet[18] = WIDEVANE[lookupByteMapIndex(WIDEVANE_MAP, 7, settings.wideVane)];
+    packet[7] += CONTROL_PACKET_2[0];
+  }
+  // add the checksum
+  byte chkSum = checkSum(packet, 21);
+  packet[21] = chkSum;
+}
+
+bool P090_data_struct::update(const heatpumpSettings& settings) {
+  while(!canSend(false)) { delay(10); }
+
+  byte packet[PACKET_LEN] = {};
+  createPacket(packet, settings);
+  writePacket(packet, PACKET_LEN);
+
+  while(!canRead()) { delay(10); }
```

Comment:
This is also blocking code, waiting for up-to 10 seconds.
It also looks like it may wait forever as the status of `_waitForRead` will not change from false to true while waiting.